From a89d3a2cf2ee6e310a3ebdeb32d69521a5d8a692 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Thu, 31 Mar 2022 05:25:56 +0300
Subject: [PATCH] [SW-66675] rom: tpm: add tpm extend routine

This patch ports the TPM_Extend() routine to rom from preboot.
This api would be used to store the ppboot sha as a digest in
rom.

Change-Id: I6aff2259671aada20dc4845693a519bbc24292d6
---

diff --git a/arch/gaudi3/arc/tpm2.c b/arch/gaudi3/arc/tpm2.c
index 86cae1f..382b780 100644
--- a/arch/gaudi3/arc/tpm2.c
+++ b/arch/gaudi3/arc/tpm2.c
@@ -14,10 +14,15 @@
 #include <linux/types.h>
 #include <tpm2.h>
 #include <hl_rsa_arc.h>
+#include <stdio.h>
+#include <stdarg.h>
 
 /* Internal error of TPM command library */
 #define TPM_LIB_ERROR ((uint32_t)~0u)
 #define le32_to_cpu(x) ((__uint32_t)(__le32)(x))
+
+/* Internal error of TPM command library */
+#define TPM_LIB_ERROR ((uint32_t)~0u)
 
 static struct spi_ctx tpm_spi = {0};
 static int locality;
@@ -652,3 +657,103 @@
 
 	return 0;
 }
+
+static int tpm2_pack_byte_string(uint8_t *str, size_t size, const char *format, ...)
+{
+	va_list args;
+	size_t offset = 0, length = 0;
+	uint8_t *data = NULL;
+	uint32_t value = 0;
+
+	va_start(args, format);
+	for (; *format; format++) {
+		switch (*format) {
+		case 'b':
+			offset = va_arg(args, size_t);
+			value = va_arg(args, int);
+			length = 1;
+			break;
+		case 'w':
+			offset = va_arg(args, size_t);
+			value = va_arg(args, int);
+			length = 2;
+			break;
+		case 'd':
+			offset = va_arg(args, size_t);
+			value = va_arg(args, uint32_t);
+			length = 4;
+			break;
+		case 's':
+			offset = va_arg(args, size_t);
+			data = va_arg(args, uint8_t *);
+			length = va_arg(args, uint32_t);
+			break;
+		default:
+			va_end(args);
+			return -1;
+		}
+
+		if (offset + length > size) {
+			va_end(args);
+			return -1;
+		}
+
+		switch (*format) {
+		case 'b':
+			str[offset] = value;
+			break;
+		case 'w':
+			put_unaligned_be16(value, str + offset);
+			break;
+		case 'd':
+			put_unaligned_be32(value, str + offset);
+			break;
+		case 's':
+			memcpy(str + offset, data, length);
+			break;
+		}
+	}
+	va_end(args);
+
+	return 0;
+}
+
+int tpm2_pcr_extend(uint32_t index, uint32_t algorithm, const uint8_t *digest, uint32_t digest_len)
+{
+	/* Length of the message header, up to start of digest */
+	uint32_t offset = 33;
+	uint8_t command_v2[COMMAND_BUFFER_SIZE] = {
+		tpm_u16(TPM2_ST_SESSIONS),	/* TAG */
+		tpm_u32(offset + digest_len),	/* Length */
+		tpm_u32(TPM2_CC_PCR_EXTEND),	/* Command code */
+
+		/* HANDLE */
+		tpm_u32(index),			/* Handle (PCR Index) */
+
+		/* AUTH_SESSION */
+		tpm_u32(9),			/* Authorization size */
+		tpm_u32(TPM2_RS_PW),		/* Session handle */
+		tpm_u16(0),			/* Size of <nonce> */
+						/* <nonce> (if any) */
+		0,				/* Attributes: Cont/Excl/Rst */
+		tpm_u16(0),			/* Size of <hmac/password> */
+						/* <hmac/password> (if any) */
+
+		/* hashes */
+		tpm_u32(1),			/* Count (number of hashes) */
+		tpm_u16(algorithm),	/* Algorithm of the hash */
+		/* STRING(digest)		   Digest */
+	};
+	int ret;
+
+	/*
+	 * Fill the command structure starting from the first buffer:
+	 *     - the digest
+	 */
+	ret = tpm2_pack_byte_string(command_v2, sizeof(command_v2), "s",
+			       offset, digest, digest_len);
+	if (ret)
+		return TPM_LIB_ERROR;
+
+	return tpm_spi_sendrecv_command(command_v2, NULL, NULL);
+}
diff --git a/common/rsa/hl_rsa_arc.h b/common/rsa/hl_rsa_arc.h
index 1c71ede..7a396b5 100644
--- a/common/rsa/hl_rsa_arc.h
+++ b/common/rsa/hl_rsa_arc.h
@@ -9,6 +9,8 @@
 
 #define put_unaligned_be32(a, b) (*(uint32_t *)(b) = __builtin_bswap32(a))
 
+#define put_unaligned_be16(a, b) (*(uint16_t *)(b) = __builtin_bswap16(a))
+
 static inline uint32_t get_unaligned_be32(const void *ptr)
 {
 	const uint8_t *p = (const uint8_t *)ptr;
